protocol/x/affiliates/keeper/keeper.go

```diff
@@ -41,3 +49,154 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
 }
 
 func (k Keeper) InitializeForGenesis(ctx sdk.Context) {}
+
+func (k Keeper) RegisterAffiliate(
+	ctx sdk.Context,
+	referee string,
+	affiliateAddr string,
+) error {
+	referredByPrefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), []byte(types.ReferredByKeyPrefix))
+	if referredByPrefixStore.Has([]byte(referee)) {
+		return errors.New("Referee already exists for address: " + referee)
+	}
+	referredByPrefixStore.Set([]byte(referee), []byte(affiliateAddr))
+	return nil
+}
+
+func (k Keeper) GetReferredBy(ctx sdk.Context, referee string) (string, bool) {
+	referredByPrefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), []byte(types.ReferredByKeyPrefix))
+	if !referredByPrefixStore.Has([]byte(referee)) {
+		return "", false
+	}
+	return string(referredByPrefixStore.Get([]byte(referee))), true
+}
+
+func (k Keeper) AddReferredVolume(
+	ctx sdk.Context,
+	affiliateAddr string,
+	referredVolumeFromBlock dtypes.SerializableInt,
+) error {
+	affiliateReferredVolumePrefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), []byte(types.ReferredVolumeKeyPrefix))
+	var referredVolume dtypes.SerializableInt
+	if !affiliateReferredVolumePrefixStore.Has([]byte(affiliateAddr)) {
+		referredVolume = dtypes.NewInt(0)
+	} else {
+		err := referredVolume.Unmarshal(affiliateReferredVolumePrefixStore.Get([]byte(affiliateAddr)))
+		if err != nil {
+			return err
+		}
+	}
+	referredVolumeBigInt := referredVolume.BigInt()
+	referredVolumeBigInt.Add(referredVolumeBigInt, referredVolumeFromBlock.BigInt())
+	updatedReferedVolume := dtypes.NewIntFromBigInt(referredVolumeBigInt)
+
+	updatedReferredVolumeBytes, err := updatedReferedVolume.Marshal()
+	if err != nil {
+		return errors.New("Error marshalling referred volume")
+	}
+	affiliateReferredVolumePrefixStore.Set([]byte(affiliateAddr), updatedReferredVolumeBytes)
+	return nil
+}
+
+func (k Keeper) GetReferredVolume(ctx sdk.Context, affiliateAddr string) (dtypes.SerializableInt, bool) {
```

Comment:
Also return `*big.Int` similar to above.

Reply:
Change this signature to return `(dtypes.SerializableInt, error)` since error is returned below
Don't think we need to differentiate between `existing affiliate referring 0 volume` vs. nonexisting affiliate. This way you also don't need [this](https://github.com/dydxprotocol/v4-chain/pull/2141/files#diff-c03e538e2775e655ea21348c318ab5275016f108f12b3092e97bd6136e312a8fR168)